codalab/lib/bundle_cli.py

```diff
@@ -1170,26 +1185,134 @@ def derive_bundle(self, bundle_type, command, targets, metadata):
 
     @Commands.command(
         'run',
-        help='Create a bundle by running a program bundle on an input bundle.',
+        help='Create a bundle by running a program bundle on an input bundle. If local mode is specified, simulate a run bundle locally, producing bundle contents in the local environment',
         arguments=(
             Commands.Argument('target_spec', help=ALIASED_TARGET_SPEC_FORMAT, nargs='*', completer=TargetsCompleter),
             Commands.Argument('command', metavar='[---] command', help='Arbitrary Linux command to execute.', completer=NullCompleter),
             Commands.Argument('-w', '--worksheet-spec', help='Operate on this worksheet (%s).' % WORKSHEET_SPEC_FORMAT, completer=WorksheetsCompleter),
+            Commands.Argument('--local', action='store_true', help='Simulate a run bundle locally.'),
         ) + Commands.metadata_arguments([RunBundle]) + EDIT_ARGUMENTS + WAIT_ARGUMENTS,
     )
     def do_run_command(self, args):
         client, worksheet_uuid = self.parse_client_worksheet_uuid(args.worksheet_spec)
         args.target_spec, args.command = cli_util.desugar_command(args.target_spec, args.command)
-        targets = self.parse_key_targets(client, worksheet_uuid, args.target_spec)
         metadata = self.get_missing_metadata(RunBundle, args)
-        new_bundle = client.create(
-            'bundles',
-            self.derive_bundle(RunBundle.BUNDLE_TYPE, args.command, targets, metadata),
-            params={'worksheet': worksheet_uuid},
+
+        if args.local:
+            docker_image = metadata.get('request_docker_image', None)
+            if docker_image is None:
+                raise UsageError('--request-docker-image [docker-image] must be specified when running in local mode')
+
+            uuid = generate_uuid()
+            bundle_path = '/opt/{}'.format(uuid)
+            command = args.command
+            request_network = None
+            dependencies = [
+                (u'{}'.format(key), u'/{}_dependencies/{}'.format(uuid, key)) for key, target in self.parse_target_specs(args.target_spec)
+            ]
+
+            # Set up a directory to store the bundle.
+            remove_path(bundle_path)
+            os.mkdir(bundle_path)
+
+            for dependency_path, docker_dependency_path in dependencies:
+                child_path = os.path.join(bundle_path, dependency_path)
+                os.symlink(docker_dependency_path, child_path)
+
+            dc = DockerClient()
+            container_id = dc.start_container(bundle_path, uuid, command, docker_image, request_network, dependencies)
+            print >>self.stdout, '===='
```

Comment:
This information is printed out pretty much at the same time the container is started. So it is effectively printed before we run the job.

Furthermore, we cannot get the container id before we start the container.